EIPS/eip-6224.md

```diff
@@ -0,0 +1,244 @@
+
+   /**
+    *  @notice The function that gets the current dependency injector
+    *  @return the current dependency injector
+    */
+   function getInjector() external view returns (address);
+}
+```
+
+- The `Dependant` contract MUST pull its dependencies in the `setDependencies` method from the passed `contractsRegistry` address.
+- The `Dependant` contract MAY store the dependency injector address in the special slot `0x3d1f25f1ac447e55e7fec744471c4dab1c6a2b6ffb897825f9ea3d2e8c9be583` (obtained as `bytes32(uint256(keccak256("eip6224.dependant.slot")) - 1)`).
+
+
+## Rationale
+
+The EIP proposes a way to manage the system's smart contracts and maintain a clear usage of their dependencies. There are a few design decisions that have to be specified explicitly:
+
+### ContractsRegistry Rationale
+
+**Usage**
+
+The `ContractsRegistry` SHOULD be the topmost smart contract in the system. It SHOULD NOT depend on other contracts. The extensions of this EIP SHOULD add proper access control check to the described non-view methods. 
+
+The `getContract` and `getImplementation` methods SHOULD revert if the nonexisting contracts are asked. This logic MAY be used to ensure that proper dependencies are provided to the dependants.
+
+The `ContractsRegistry` MAY be set behind the proxy to enable runtime addition of custom methods. Applications MAY also leverage the pattern to develop custom tree-like `ContractsRegistry` data structures.
+
+**Contracts identifier**
+
+The `string` contracts identifier is chosen over the `uint256` and `bytes32` to maintain code readability and reduce the human-error chances when interacting with the `ContractsRegistry`. Being the topmost smart contract, it MAY be typical for the users to interact with it via block explorers or DAOs. Clarity, not gas usage, is a factor of choice.
+
+**Proxy**
+
+The `Transparent` proxy is chosen over the `UUPS` proxy to hand the upgradeability responsibility to the `ContractsRegistry` itself. The extensions of this EIP MAY use the proxy of their choice.
+
+### Dependant Rationale
+
+**Dependencies**
+
+The required dependencies MUST be set in the overridden `setDependencies` method, not in the `constructor` or `initializer` methods.
+
+**Injector**
+
+Only the injector MUST be able to call the `setDependencies` and `setInjector` methods. The initial injector will be a zero address, in that case, the call MUST NOT revert on access control checks. The `setInjector` function is made `public` to support the dependency injection mechanism for factory-made contracts. However, the method SHOULD be used with extra care.
+
+The injector address MAY be stored in the dedicated slot `0x3d1f25f1ac447e55e7fec744471c4dab1c6a2b6ffb897825f9ea3d2e8c9be583` to exclude the chances of storage collision.
+
+## Reference Implementation
+
+*0xdistributedlab-solidity-library dev-modules* provides a reference implementation.
+
+- MIT licensed, no constraints on the usage.
```

Comment:
Must be CC0, since MIT *does* have usage restrictions (you must include a copy of the license).

Reply:
Removed the license mention. Should I still include a copy of the license?